### PR TITLE
Add initial version of hash function slides

### PR DESCRIPTION
**Pull request by kmschuler**
# Description
* Addition of first version of tex file for lecture "Introduction to Cryptography: Hash Functions" including the respective images used and figures created.
* Slides re-use already existing asset ../assets/images/schaer_berentsen_cover
* Known issue: Team slide yet missing
* No config changes, no new packages

## Type of Change

- [x ] Major feature (New files or assets, large changes to files)

# Checklist:

- [x] I have performed a self-review of my own changes
- [x] I have made sure that any changed LaTeX files compile properly
- [x] I have assigned at least one reviewer

### CODE REVIEW

File: slides/hash_functions.tex [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+		\begin{figure}[h]
+  			\center
+			\input{../assets/figures/IBAN_checksum.tex}
+		\end{figure}
+
+\footnotesize Steps:
+	\begin{itemize}
+		\item Take BBAN and append Country Code and 00 (empty checksum)
+		\item Translate characters of country code into 9 +position in alphabet
```
Comment (golmit) [inline slides/hash_functions.tex:63]: Could we adjust the spacing for the calculation? 
```suggestion
		\item Translate characters of country code into 9 + position in alphabet
```

File: slides/hash_functions.tex [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+		\item Perform mod97 operation and subtract from 98 to arrive at checksum
+		\item I.e., 98 - 4835012345678009121700 mod 97 = 56
```
Comment (golmit) [inline slides/hash_functions.tex:65]: Is 'e.g.' more appropriate?

File: slides/hash_functions.tex [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+\uncover<1->{In Bitcoin context, the functions \color{focus}SHA2.256 \color{black} (in short SHA256) and \color{focus}RIPEMD160 \color{black}are used. Both satisfy the above criteria.}
+	
+\end{frame}
+%%%	
+
+%%%
+\begin{frame}{Avalanche Effect with SHA2}
+
+\begin{center}
+$SHA256(<pre-image>)$
```
Comment (golmit) [inline slides/hash_functions.tex:116]: Maybe don't use math notation for <pre-image>? Instead use italic text?
```suggestion
$SHA256(<$\textit{pre-image}$>)$
```

File: slides/hash_functions.tex [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+\normalsize}
+
+	
+\end{frame}
+%%%	
+
+%%%
+\begin{frame}{Example in Bitcoin Context}
+
+How much time would it take to find a pre-image that generates a specific hash using the cryptographic hash functions applied in Bitcoin with a try-and-error approach?
```
Comment (golmit) [inline slides/hash_functions.tex:175]:
```suggestion
How much time would it take to find a pre-image that generates a specific hash using the cryptographic hash functions applied in Bitcoin with a trial and error approach?
```

File: slides/hash_functions.tex (hunk 2 above)
Comment (Chainomics) [reply slides/hash_functions.tex:65]: Yes

File: slides/hash_functions.tex [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+Deterministic algorithm (\color{focus}function, $H()$\color{black}) that maps data of quasi-arbitrary size (\color{focus}pre-image, $m$\color{black}) to fixed-length bit string (\color{focus}hash value, $h$\color{black}).
+
+	\begin{align}
+		h = H(m)
+		\label{eq:hash_function}
+	\end{align}
+	\vspace{1em}
+	
+\uncover<1->{Example of a simplified hash function:
```
Comment (Chainomics) [inline slides/hash_functions.tex:34]: Before we go into the example, we should quickly introduce mod operations. Let's use the mod wheel from the book / original lecture and show some simple examples.

File: slides/hash_functions.tex [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+\vspace{0.5em}
+\tiny 
+$P(h \leq \color{focus}0\color{black}FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF) = \dfrac{1}{16}$\\
+\vspace{1em}
+$P(h \leq \color{focus}00\color{black}FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF) = \left( \dfrac{1}{16} \right)^{2} $
+
+\normalsize}
+
+	
+\end{frame}
```
Comment (Chainomics) [inline slides/hash_functions.tex:169]: We could give the students the mining.R script, and let them experiment with it. @matnad  I'd appreciate it, if you could quickly code that in Python. We could always use the R script, but Python seems to be much better suited for these kinds of examples.

File: slides/hash_functions.tex [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+%%%	
+
+%%%
+\begin{frame}{Example in Bitcoin Context}
```
Comment (Chainomics) [inline slides/hash_functions.tex:173]: I am afraid the example is not correct in this context. In the original lecture, it has been used in the context of Bitcoin addresses. While it is correct that Bitcoin addresses also use (multiple) hash functions,  the derivation process also involves EC math. Please remove the slide.

**Review by Chainomics — COMMENTED** [review]
Thanks @kmschuler. I have accepted @golmit proposals and added another commit with some changes on top of it. However, there are some remaining issues. Please check inline comments for detailed feedback.

File: slides/hash_functions.tex (hunk 6 above)
Comment (matnad) [reply slides/hash_functions.tex:169]: Implemented in 6ec285b

File: slides/hash_functions.tex (hunk 5 above)
Comment (kmschuler) [reply slides/hash_functions.tex]: Proposal for implementation included in commit #99343a9

File: slides/hash_functions.tex (hunk 6 above)
Comment (kmschuler) [reply slides/hash_functions.tex:208]: How (if) is this to be integrated in the slides?

Comment (Chainomics) [reply slides/hash_functions.tex:208]: I usually use a placeholder slide that describes the general idea and contains a repo link to the script. Please remove my name from the example and use `my_miner("This is an example", leading_zeros=5)` instead.

Comment (Chainomics) [reply slides/hash_functions.tex:208]: So it is mainly for the video and for students who want to test it themselves.

**Review by Rothj — COMMENTED** [review]
I saw some inconsistencies in notation of modular arithmetic in our slides. Open to discuss if we want to change the notation style, but we should definitely use it consistently.

File: slides/hash_functions.tex (hunk 6 above)
Comment (kmschuler) [reply slides/hash_functions.tex:209]: Background slide to the script added in commit 5f8c5f6
@matnad, the link to the script is yet empty.

File: slides/hash_functions.tex [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+		\end{figure}
+	\end{minipage}
+
+
+
+\uncover<1->{
+\vspace{0.5em}
+\textbf{Simplified hash function}
+	\begin{align}
+		h = m \: \mod \: 12
```
[the remark this answers is not in the document]
Comment (kmschuler) [reply slides/hash_functions.tex:77]: I'm not sure I understand, @Rothj .
According to the modulo you're suggesting in the other examples, I'd now see this as h \ (mod\ 12) = m, not?

File: slides/hash_functions.tex [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+  			\center
+			\input{../assets/figures/IBAN_checksum.tex}
+		\end{figure}
+
+\footnotesize Steps:
+	\begin{itemize}
+		\item Take BBAN and append Country Code and 00 (empty checksum).
+		\item Translate characters of country code into 9 + \texttt{<position in alphabet>}.
+		\item Perform mod97 operation and subtract from 98 to arrive at checksum.
+		\item e.g., $98 - 4835012345678009121700 \mod 97 = 56$.
```
[the remark this answers is not in the document]
Comment (kmschuler) [reply slides/hash_functions.tex]: Adopted in commit 14b135f

**Review by Chainomics — APPROVED** [review]
Great slide deck! Thanks for the changes @kmschuler!

**Pull request merged**